openml/runs/functions.py

```diff
@@ -540,40 +563,67 @@ def _run_model_on_fold(model, task, rep_no, fold_no, sample_no, can_measure_runt
         model : sklearn model
             The model trained on this fold
     """
+
     def _prediction_to_probabilities(y, model_classes):
         # y: list or numpy array of predictions
-        # model_classes: sklearn classifier mapping from original array id to prediction index id
+        # model_classes: sklearn classifier mapping from original array id to
+        # prediction index id
         if not isinstance(model_classes, list):
-            raise ValueError('please convert model classes to list prior to calling this fn')
+            raise ValueError('please convert model classes to list prior to '
+                             'calling this fn')
         result = np.zeros((len(y), len(model_classes)), dtype=np.float32)
         for obs, prediction_idx in enumerate(y):
             array_idx = model_classes.index(prediction_idx)
             result[obs][array_idx] = 1.0
         return result
 
-    # TODO: if possible, give a warning if model is already fitted (acceptable in case of custom experimentation,
+    # TODO: if possible, give a warning if model is already fitted (acceptable
+    # in case of custom experimentation,
     # but not desirable if we want to upload to OpenML).
 
-    train_indices, test_indices = task.get_train_test_split_indices(repeat=rep_no,
-                                                                    fold=fold_no,
-                                                                    sample=sample_no)
+    train_indices, test_indices = task.get_train_test_split_indices(
+        repeat=rep_no, fold=fold_no, sample=sample_no)
+    if task.task_type in (
+            "Supervised Classification",
+            "Supervised Regression",
+            "Learning Curve",
+    ):
+        X, Y = task.get_X_and_y()
+        trainX = X[train_indices]
+        trainY = Y[train_indices]
+        testX = X[test_indices]
+        testY = Y[test_indices]
+    elif task.task_type in (
+            "Clustering",
+    ):
+        trainX = train_indices
+        testX = test_indices
+    else:
+        raise NotImplementedError(task.task_type)
 
-    X, Y = task.get_X_and_y()
-    trainX = X[train_indices]
-    trainY = Y[train_indices]
-    testX = X[test_indices]
-    testY = Y[test_indices]
     user_defined_measures = collections.OrderedDict()
 
     try:
         # for measuring runtime. Only available since Python 3.3
         if can_measure_runtime:
             modelfit_starttime = time.process_time()
-        model.fit(trainX, trainY)
+
+        if task.task_type in (
```

Comment:
I would actually suggest to create somewhere a mapping, e.g., in `openml.tasks.task.py`

```
class TaskTypeEnum(object):
  SUPERVISED_CLASSIFICATION = 1
  SUPERVISED_REGRESSION = 2
  LEARNING_CURVE = 3
  # etc
```

And then refer to the task.task_type_id with `TaskTypeEnum.SUPERVISED_CLASSIFICATION`, etc

Reply:
That won't work because of python2 (actually, could you ask Andy in person about dropping that? ;)) But the idea is great!

Reply:
why wouldn't that work? Note that (despite the name) this is not an enum, but a conventional class.

Reply:
Any resolution on whether you guys want this done? 
I think Matthias was pointing out that Enums are quite new in Python:
https://docs.python.org/3/library/enum.html
IIUC, You have to use other libraries to get enum's in Python 2:
https://stackoverflow.com/questions/36932/how-can-i-represent-an-enum-in-python

So you'd be stuck with not perfectly compatible code.

Also, if I'm right you want me to replace this by something like 
```
if task.task_type_id in (TaskTypeEnum.SUPERVISED_CLASSIFICATION, TaskTypeEnum.LEARNING_CURVE)
```
Right? Is that what you want?

Let me know what you think. I wouldn't have time for this in the next week, I'm afraid. Also, where do you want me to put this enum class so that it's available everywhere? The more I think about it, that would be a change that affects more than just the regression PR, so it may be best to have a separate PR for this?

Reply:
Great. Can we merge this PR then and open a new one for the enum refactoring? Unless there is anything else?

Reply:
Who is going to make the follow up pr?

Reply:
I opened a new issue here: 
https://github.com/openml/openml-python/issues/585

I'll ask Pieter and Bilge if they can pitch in to speed things up. Otherwise, I can also do it as soon as I have some time.

Shall we merge this PR before we start working on the refactoring? Or do you want to keep this open until the other PR is merged and I have time to refactor this PR again?

Reply:
> Shall we merge this PR before we start working on the refactoring? Or do you want to keep this open until the other PR is merged and I have time to refactor this PR again?

My personal preference would be to first merge the other PR. It should not be a lot of work, and it will make the code much less bug prone and more readable. Previous experience has made me a little bit afraid that if we merge this first, the other PR will probably be forgotten and wait until we get the same discussion around a new task type. 

Disadvantage of making a other PR first would be that this PR will get some merge conflicts that need to be resolved. I am also fine with having both options in one PR, to avoid the merge conflicts. @mfeurer what do you think? 

I know that it is a bit frustrating because for the previous task types we did not immediately pick up on this. However, any help on this would be greatly appreciated and contribute greatly to a more healthy code base. :)

Reply:
I kind of agree with @janvanrijn. It would be great if you could do this prior to continue with this task and the rebase this one. Maybe @PGijsbers and/or @Bilgecelik can help you with this as you proposed?

Reply:
Yes I can work on it.

Reply:
Bilge, if you need any support, let me know!